src/software/geom/algorithms/end_in_obstacle_sample.cpp

```diff
@@ -0,0 +1,88 @@
+#include "software/geom/algorithms/end_in_obstacle_sample.h"
+
+#include <random>
+
+#include "software/ai/navigator/obstacle/obstacle.hpp"
+#include "software/geom/point.h"
+
+
+std::optional<Point> endInObstacleSample(const std::vector<ObstaclePtr> obstacles,
+                                         const Point &point,
+                                         const Rectangle &navigable_area,
+                                         int initial_count, double rad_step,
+                                         int per_rad_step, double range)
+{
+    bool point_in_obstacle = false;
+    for (auto const &obstacle : obstacles)
+    {
+        if (obstacle->contains(point) || !contains(navigable_area, point))
+        {
+            if (!point_in_obstacle)
+            {
+                point_in_obstacle = true;
+            }
+
+            Point closest_point            = obstacle->closestPoint(point);
+            bool closest_point_in_obstacle = false;
+            for (auto const &obstacle : obstacles)
+            {
+                if (obstacle->contains(closest_point))
+                {
+                    closest_point_in_obstacle = true;
+                    break;
+                }
+                if (!contains(navigable_area, closest_point))
+                {
+                    closest_point_in_obstacle = true;
+                    break;
+                }
+            }
+
+            if (closest_point_in_obstacle)
+            {
+                break;
+            }
+            else
+            {
+                return closest_point;
```

Comment:
`closestPoint` helper functions return the closest point on the edge of the obstacle, ideally we would shift the point slightly outside of the obstacle just to make sure the destination is not thought to be `contain`ed within an obstacle (I think most `contains` functions count the edge of an obstacle as inside of it)

One way you could do this is to shift whatever value returned by this function (considering the multiple `return` statements). You could shift it by doing something like 
```cpp
closest_point += (closest_point - point).normalize(some_small_margin_like_MILLIMETERS_PER_METER);
```

Reply:
make a new constant inside constants.h ?

Reply:
Since it's not a shared constant with other files, you can probably just have it here. In the case of this one, it might be a good idea to have a function argument when has `OBSTACLE_AVOIDANCE_BUFFER_CENTIMETERS` as a default value.